### PR TITLE
[SDL-0238] Keyboard Enhancements

### LINKED ISSUE #340: [SDL 0238] Keyboard Enhancements
Proposal: [Keyboard Enhancements](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0238-Keyboard-Enhancements.md)

> This proposal is about adding a new keyboard layout and adding new properties for keyboards supported by SDL. The following keyboard enhancements are discussed in this proposal.

Enhancement #1: Support for Numeric Keyboard.

Enhancement #2: Allow apps to mask entered characters.

Enhancement #3: Allow apps to change special characters shown on the keyboard layout.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/754

Steering Committee Decision:

>The Steering Committee voted to accept this proposal with the revisions outlined in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/754#issuecomment-511379464) from the author.  The project maintainer also agreed to the author's point regarding `INVALID_DATA` (12).

> The proposal .md file was updated to reflect these revisions on 7/22/19.

### PR DESCRIPTION
**Pull request by ymalovanyi**
Fixes #340

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior

#### Unit Tests
Added unit tests cover [[SDL-0238] Keyboard Enhancements](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0238-Keyboard-Enhancements.md) changes

#### Core Tests
* each of the properties in the struct `KeyboardCapabilities` can be received from HMI
* each of the values in the enum `KeyboardInputMask` can be received from HMI
* `KeyboardEvent` enum contains two additional values `"INPUT_KEY_MASK_ENABLED"`, and `"INPUT_KEY_MASK_DISABLED"`
* each of the properties in the struct `ConfigurableKeyboards` can be received from HMI
* `keyboardCapabilities` property is present in `WindowCapability` struct, and contains `configurableKeys` property
* `KeyboardLayout` enum contains additional value `"NUMERIC"`
* `maskInputCharacters` and `customizeKeys` properties in the struct `KeyboardProperties` can be received from HMI
* in `SetGlobalProperties` message request `keyboardProperties` parameter can be sent with `keyboardLayout` property set to `"NUMERIC"`, and defined `maskInputCharacters` and `customizeKeys` properties

Core version / branch / commit hash / module tested against: [LuxoftSDL/sdl_core#sdl_0238_keyboard_enhancements](https://github.com/LuxoftSDL/sdl_core/tree/feature/sdl_0238_keyboard_enhancements)
HMI name / version / branch / commit hash / module tested against: [LuxoftSDL/sdl_hmi#sdl_0238_keyboard_enhancements](https://github.com/LuxoftSDL/sdl_hmi/tree/feature/sdl_0238_keyboard_enhancements)

### Summary
Applied [[SDL-0238] Keyboard Enhancements](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0238-Keyboard-Enhancements.md) changes

##### Breaking Changes
* `N/A`

##### Enhancements
* New structs `KeyboardCapabilities`, `ConfigurableKeyboards`
* New enum `KeyboardInputMask`
* Added required getters/setters in related RPC classes

##### Bug Fixes
* `N/A`

### Tasks Remaining:
* `N/A`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by ymalovanyi** [comment]
@santhanamk the PR is ready for Ford review.

**Review by santhanamk — APPROVED** [review]
@ymalovanyi Code looks good.

**Comment by jordynmackool** [comment]
@ymalovanyi is this ready for Livio Reivew?

**Comment by vladmu** [comment]
@jordynmackool @crokita the description was fixed to include all sections from the template and the code was tested against the Core and HMI (corresponded links also included in the description). Please review. Also, please take into account that the JavaScript Suite doesn't contain Choice Set Manager (it seems it will be added in the implementation of https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0272-sdl-javascript-manager-layer.md) therefore only RPC changes of the proposal included here and we have tested this against Core with cases defined in the description.

**Comment by jordynmackool** [comment]
@ymalovanyi / @vladmu the proposal markdown file has been updated per the revisions included in the accepted with revisions review issue: [Revise SDL-0238 Keyboard Enhancements](https://github.com/smartdevicelink/sdl_evolution/pull/1119).

Please make the needed updates to this PR and then tag @renonick87 to review when ready. Thanks!

**Comment by BogMW** [comment]
@renonick87 all required adjustments were made, the PR is ready for review.

**Review by renonick87 — CHANGES_REQUESTED** [review]
@ymalovanyi  From an initial review, the content of this PR looks good. But with the addition of the ChoiceSetManager in https://github.com/smartdevicelink/sdl_javascript_suite/pull/371, there are additional updates needed for the manager to support this feature.

**Comment by vladmu** [comment]
@renonick87, we've made additional changes to the ChoiceSetManager. Moreover, this aligned with the last state of Java PR https://github.com/smartdevicelink/sdl_java_suite/pull/1587. We have tested the code with the Core/HMI. Please review.

File: lib/js/src/manager/screen/choiceset/_PresentChoiceSetOperation.js [hunk 1]
```diff
@@ -306,6 +306,8 @@ class _PresentChoiceSetOperation extends _Task {
             } else if (onKeyboardInput.getEvent() === KeyboardEvent.ENTRY_ABORTED || onKeyboardInput.getEvent() === KeyboardEvent.ENTRY_CANCELLED) {
                 // notify of abort / cancelation
                 this._keyboardListener.onKeyboardDidAbortWithReason(onKeyboardInput.getEvent());
+            } else if (onKeyboardInput.getEvent().equals(KeyboardEvent.INPUT_KEY_MASK_ENABLED) || onKeyboardInput.getEvent().equals(KeyboardEvent.INPUT_KEY_MASK_DISABLED)) {
```
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/_PresentChoiceSetOperation.js:309]: This addition should also be included in the _PresentKeyboardOperation.js file.

**Review by renonick87 — COMMENTED** [review]
@ymalovanyi See my feedback for a minor suggestion.

**Comment by BogMW** [comment]
Hey @renonick87 , changes were added.

**Review by renonick87 — APPROVED** [review]

**Pull request merged**